### PR TITLE
Add other dp events, and fix possible infinite digest loop

### PR DESCRIPTION
**Pull request by CapitanMorgan**
Adds the previously unimplemented items from http://eonasdan.github.io/bootstrap-datetimepicker/Events/

Fixes digest loop i mentioned in #26  and by doing so allows for the functionality desired in #26

### CODE REVIEW

**Comment by atais** [comment]
@CapitanMorgan whoa, why the tests failed so badly ?

**Comment by CapitanMorgan** [comment]
When i check out master locally, it was already broken. So I started about trying to fix it.

Edit: it appeared to be mainly due to issues with how i was running it , not sure if its because of windows or git bash. Eventually once i got it running i was able to fix them to work with the new changes locally. I still need to figure out how if possible to add something to test that the infinite digest loop doesn't still occur. At minimum i want to make a js fiddle showing what was happening when you use certain settings with the prior version.

**Comment by atais** [comment]
You are right. So at the time of last commit to master Travis was OK, but actually, now it does not pass.
https://travis-ci.org/atais/angular-eonasdan-datetimepicker/builds/197936266

So I think 1st thing would be to sort this out and later I can think of adding new stuff.

FFS, JavaScript world could not be stable for even couple of months :angry:

**Comment by atais** [comment]
I fixed the tests in #29

File: examples/double.html [hunk 1]
```diff
@@ -25,31 +25,41 @@
             <div class="col-md-3">
                 <div class="form-group">
                     <div class="input-group" datetimepicker
-                         on-change="vm.optionsTo.minDate = vm.dateFrom"
+                         on-change="vm.onChangeDatePickerFrom()"
```
Comment (atais) [inline examples/double.html:28]: Please keep the old tests untouched.

1.  All new features are optional, so I believe the old scenarios should work?
2. Please provide NEW tests for new features

File: package.json [hunk 2]
```diff
@@ -23,9 +23,7 @@
   "scripts": {
     "pretest": "npm install",
     "test": "(npm start > /dev/null &) && (gulp protractor)",
-
```
Comment (atais) [inline package.json:26]: Those changes are probably not necessary as well.

File: dist/angular-eonasdan-datetimepicker.js [hunk 3]
```diff
@@ -39,22 +43,60 @@
...
                     var isDateEqual = function (d1, d2) {
-                        return moment.isMoment(d1) && moment.isMoment(d2) && d1.valueOf() === d2.valueOf();
+                      if(moment.isMoment(d1) && moment.isMoment(d2)){
+                        if(d1.isValid() && d2.isValid()){
+                          return (d1.valueOf() === d2.valueOf());
+                        }
+                      } else if ((!d1 && d2 && (d1 === d1)) || (d1 && !d2 && (d2 === d2))){
+                        return false; // the d1 === d1 is to avoid NaN which is passed in iniitilization
+                      }
+                      return true; // return true for invald scenarios to prevent possible infinite digest loop
```
Comment (atais) [inline dist/angular-eonasdan-datetimepicker.js:53]: what are the _invalid_ scenarios?
and why would you return `true` on `isDateEqual` check when they are **invalid**? 

If properties are not date objects, or whatever is they reason why they can't be checked - I think- this method should return `false`.

**Review by atais — CHANGES_REQUESTED** [review]

**Comment by atais** [comment]
So ideally I would like you to create a test scenario describing this infinite loop that would fail on current master.
And from there you could create a fix for it and present the whole as a PR.

**Pull request closed without merging**